qemu/tests/block_hotplug.py

```diff
@@ -88,153 +76,81 @@ def unplug_device(vm, get_disk_cmd, device):
-                if blk_extra_param:
-                    for item in blk_extra_param.split():
-                        key, value = item.split("=", 1)
-                        device.set_param(key, value)
-                device.hotplug(vm.monitor)
-                ver_out = device.verify_hotplug("", vm.monitor)
-                if not ver_out:
-                    err = "%s is not in qtree after hotplug" % pci_type
-                    test.fail(err)
-                plug_status = utils_misc.wait_for(lambda: len(get_new_disk(disks_before_plug,
-                                                  find_disk(vm, get_disk_cmd))) != 0, pause)
-                if plug_status:
-                    disks_after_plug = find_disk(vm, get_disk_cmd)
-                    new_disks = get_new_disk(disks_before_plug, disks_after_plug)
-                else:
-                    test.fail("Can't get new disks")
-                if params.get("need_controller", "no") == "yes":
-                    info_qtree = vm.monitor.info('qtree', False)
-                    qtree = qemu_qtree.QtreeContainer()
-                    qtree.parse_info_qtree(info_qtree)
-                    for node in qtree.get_nodes():
-                        if node.qtree.get("id") == device.get_param("id"):
-                            try:
-                                controller_id = node.parent.qtree.get("id").split(".")[0]
-                            except AttributeError:
-                                test.fail("can't get parent of:\n%s" % node)
-                            controller_device_dict.setdefault(controller_id, []).append(device)
-                            break
-                    else:
-                        test.fail("Can't find device '%s' in qtree" % device.get_param("id"))
+                        test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
+
+                    status, output = session.cmd_status_output(test_cmd,
+                                                               timeout=disk_op_timeout)
+                    if status:
+                        test.fail("Check for block device failed ")
+
+                devs = [dev for dev in devs if not isinstance(dev, qdevices.QDrive)]
+                if not devs:
+                    test.fail("Incorrect number of returned devices. At least "
+                              "one drive and one device is required.")
+                device_list.extend(devs)
             else:
-                if params.get("drive_format") in pci_type:
-                    get_disk_cmd += " | egrep -v '^/dev/[hsv]da[0-9]*$'"
+                for device in vm.devices:
+                    if device.get_param("id") == img_list[num + 1]:
+                        device_list.append(device)
 
-                device.set_param("id", img_list[num + 1])
-                new_disks = find_disk(vm, get_disk_cmd)
+        plug_tag = "after_plug"
+        status = run_sub_test(params, plug_tag)
+        if not status:
```

Comment:
This works, but is quite hard to grasp. How about calling it `vm_switched_off`? It'd mean inverting the logic (`if vm_switched_off; return`) but IMO it'd be easier to follow.

The same in `run_sub_test` docstring, something like `:return: whether the machine was successfully shut-down as part of the subtest`.